test/zzz-rsyslog.sh

```diff
@@ -320,9 +322,120 @@ else
 fi
 
 # switch LOGGING_FILE_PATH to console
-oc set env daemonset/rsyslog LOGGING_FILE_PATH=console
+oc set env $rsyslog_ds LOGGING_FILE_PATH=console
 os::cmd::try_until_failure "oc get pods $rpod > /dev/null 2>&1"
 os::cmd::try_until_success "oc get pods 2> /dev/null | grep -q 'rsyslog.*Running'"
 rpod=$( get_running_pod rsyslog )
 oc logs $rpod -c rsyslog > $ARTIFACT_DIR/rsyslog.console.log 2>&1
 os::cmd::expect_failure "grep 'oc exec <pod_name> -- logs' $ARTIFACT_DIR/rsyslog.console.log"
+
+# switch back LOGGING_FILE_PATH to default
+oc set env $rsyslog_ds LOGGING_FILE_PATH-
+os::cmd::try_until_failure "oc get pods $rpod > /dev/null 2>&1"
+os::cmd::try_until_success "oc get pods 2> /dev/null | grep -q 'rsyslog.*Running'"
+rpod=$( get_running_pod rsyslog )
+
+artifact_log "************************"
+artifact_log "Testing undefined fields"
+artifact_log "************************"
+
+artifact_log "0. Preparing test data"
+cfgdir=$( mktemp -d )
+oc extract cm/rsyslog --to=$cfgdir
+cat > $cfgdir/66-debug.conf << EOF
+set $.ret = parse_json("{\"undefined1\":\"undefined1\",\"undefined11\":1111,\"undefined12\":\"True\",\"empty1\":\"\",\"undefined2\":{\"undefined2\":\"undefined2\",\"\":\"\",\"undefined22\":2222,\"undefined23\":\"False\"},\"undefined3\":{\"emptyvalue\":\"\"},\"undefined4\":\"undefined4\",\"undefined5\":\"undefined5\",\"undefined.6\":\"undefined6\"}", "\$!");
```

Comment:
I'm running https://github.com/rsyslog/rsyslog/blob/master/tests/rscript_parse_json.sh#L11 - I modified it to add a boolean and an array like this:
```
 local4.* {
-       set $.ret = parse_json("{ \"c1\":\"data\" }", "\$!parsed");
+       set $.ret = parse_json("{ \"c1\":\"data\",\"bool\":True,\"ary\":[1,2,3] }", "\$!parsed");
        action(type="omfile" file=`echo $RSYSLOG_OUT_LOG` template="outfmt")
```
This is the output
```
{ "parsed": { "c1": "data", "bool": true, "ary": [ 1, 2, 3 ] } }
```
So it looks like you can use `\"undefined\":True` as the boolean value.

Perhaps the problem is when rsyslog passes the data to the mmexternal undefined_field program?  I'm not sure what problem you are trying to solve.

Reply:
Thanks, Rich.  When I set the following test data, then I got the errors [1].  So, first I suspected the boolean, but it'd be ok.  Then, \"\":\"\" is not acceptable?
`set $.ret = parse_json("{\"undefined1\":\"undefined1\",\"undefined11\":1111,\"undefined12\":True,\"empty1\":\"\",\"undefined2\":{\"undefined2\":\"undefined2\",\"\":\"\",\"undefined22\":2222,\"undefined23\":False},\"undefined3\":{\"emptyvalue\":\"\"},\"undefined4\":\"undefined4\",\"undefined5\":\"undefined5\",\"undefined.6\":\"undefined6\"}", "\$!");`
```
[1]
Jul 15 05:56:46 rsyslog-v25vh rsyslogd-2207: error during parsing file /etc/rsyslog.d/66-debug.conf, on or before line 1: invalid character '"' in expression - is there an invalid escape sequence somewhere? [v8.37.0-9.el7 try http://www.rsyslog.com/e/2207 ]
Jul 15 05:56:46 rsyslog-v25vh rsyslogd-2207: error during parsing file /etc/rsyslog.d/66-debug.conf, on or before line 1: invalid character '"' in expression - is there an invalid escape sequence somewhere? [v8.37.0-9.el7 try http://www.rsyslog.com/e/2207 ]
Jul 15 05:56:46 rsyslog-v25vh rsyslogd: [origin software="rsyslogd" swVersion="8.37.0-9.el7" x-pid="14" x-info="http://www.rsyslog.com"] start
Jul 15 05:56:46 rsyslog-v25vh rsyslogd-2441: internal error:  getJSONRootAndMutex; invalid indicator char {(7b) [v8.37.0-9.el7 try http://www.rsyslog.com/e/2441 ]
Jul 15 05:56:46 rsyslog-v25vh rsyslogd-2441: internal error:  getJSONRootAndMutex; invalid indicator char {(7b) [v8.37.0-9.el7 try http://www.rsyslog.com/e/2441 ]
```

Reply:
>  Then, "":"" is not acceptable?

I don't know.  I don't know what it is complaining about

Reply:
This is the output in the artifact out file when running the test:
```
set $.ret = parse_json("{\"undefined1\":\"undefined1\",\"undefined11\":1111,\"undefined12\":True,\"empty1\":\"\",\"undefined2\":{\"undefined2\":\"undefined2\",\"\":\"\",\"undefined22\":2222,\"undefined23\":\"False\"},\"undefined3\":{\"emptyvalue\":\"\"},\"undefined4\":\"undefined4\",\"undefined5\":\"undefined5\",\"undefined.6\":\"undefined6\"}", "$!");
```
* I suggest to use single quotes for the string passed to `parse_json` to avoid have to escape so many double quotes https://www.rsyslog.com/doc/v8-stable/rainerscript/constant_strings.html
* the final `"$!"` looks wrong - according to https://www.rsyslog.com/doc/v8-stable/rainerscript/functions/rs-parse_json.html it should look like this - `"\$!"` - the shell here document is eating one of the escapes `\` - try using single quotes for this too:
```
cat > $cfgdir/66-debug.conf << EOF
set \$.ret = parse_json('{"undefined1":"undefined1","undefined11":1111,"undefined12":True,"empty1":"","undefined2":{"undefined2":"undefined2","":"","undefined22":2222,"undefined23":"False"},"undefined3":{"emptyvalue":""},"undefined4":"undefined4","undefined5":"undefined5","undefined.6":"undefined6"}', '\$!');
EOF
```
that should expand to 
```
set $. = .... ,  '$!');
```

Reply:
it looks like using `$.` in a bash here document is preserved, I guess because `$.` isn't a legal variable in shell, but `$!` is _not_ preserved, because `$!` has a special meaning in bash.

Reply:
Thanks, Rich.  I replaced the 66-debug.conf generation code with your suggestion.